File: python/day3.py
import sys
import re

def part1():
    pattern = r"mul\((\d+),(\d+)\)"
    matches = re.findall(pattern, lines)
    matches = [tuple(map(int, tup)) for tup in matches]
    print(matches)
    return sum(x * y for x, y in matches)

def part2():
    # Use regular expression to find all instances of "mul(number1,number2)", "do()" and "don't()"
    pattern = r"mul\((\d+),(\d+)\)|do\(\)|don't\(\)"
    matches = [(match.group(), match.start()) for match in re.finditer(pattern, lines)]
    
    # Convert matches to a list of tuples with integers and their indices
    results = []
    for match, index in matches:
        if match.startswith("mul"):
            num1, num2 = re.findall(r'\d+', match)
            results.append(((int(num1), int(num2)), index))
        else:
            results.append((match, index))
    
    res = 0
    do = True
    for val, i in results:
        if val == "don't()":
            do = False
        elif val == "do()":
            do = True
        elif do:
            res += val[0] * val[1]
    return res

if __name__ == "__main__":

    filename = sys.argv[1] if len(sys.argv) > 1 else "../data/day3.txt"
    with open(filename) as f:
        inp = f.read()

    lines = inp

    print(f"{part1()=}")
    print(f"{part2()=}")

File: cpp/day3.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <regex>

// Part 1: Sum all x*y for every 'mul(x,y)' in the input
int part1(const std::string& input) {
    std::regex mul_pattern(R"(mul\((\d+),(\d+)\))");
    std::sregex_iterator it(input.begin(), input.end(), mul_pattern);
    std::sregex_iterator end;
    int sum = 0;
    for (; it != end; ++it) {
        int x = std::stoi((*it)[1]);
        int y = std::stoi((*it)[2]);
        sum += x * y;
    }
    return sum;
}

// Part 2: Only sum x*y when in 'do' mode, toggled by 'do()' and 'don't()'
int part2(const std::string& input) {
    std::regex pattern(R"(mul\((\d+),(\d+)\)|do\(\)|don't\(\))");
    std::sregex_iterator it(input.begin(), input.end(), pattern);
    std::sregex_iterator end;
    int sum = 0;
    bool do_mode = true;
    for (; it != end; ++it) {
        if ((*it)[0] == "do()") {
            do_mode = true;
        } else if ((*it)[0] == "don't()") {
            do_mode = false;
        } else if (do_mode && (*it)[1].matched && (*it)[2].matched) {
            int x = std::stoi((*it)[1]);
            int y = std::stoi((*it)[2]);
            sum += x * y;
        }
    }
    return sum;
}

// Read the entire file as a string
std::string read_input(const std::string& filename) {
    std::ifstream infile(filename);
    std::ostringstream ss;
    ss << infile.rdbuf();
    return ss.str();
}

int main(int argc, char* argv[]) {
    std::string filename = argc > 1 ? argv[1] : "../data/day3.txt";
    std::string input = read_input(filename);
    std::cout << "part1=" << part1(input) << std::endl;
    std::cout << "part2=" << part2(input) << std::endl;
    return 0;
} 